File: src/cmds/internal_cmds.rs
//! Internal commands for tooling and bootstrap

use std::error::Error;
use clap::Subcommand;

use crate::cli::DispatchCommand;
use crate::{paths, deps};

/// Internal commands (hidden from normal help)
#[derive(Subcommand)]
pub enum InternalCmds {
    /// Bootstrap command run by bootstrap.sh
    Bootstrap,
}

impl DispatchCommand for InternalCmds {
    fn dispatch(self) -> Result<(), Box<dyn Error>> {
        match self {
            Self::Bootstrap => cmd_bootstrap(),
        }
    }
}

/// Bootstrap the Zircon installation
fn cmd_bootstrap() -> Result<(), Box<dyn Error>> {
    println!("=== Zircon Bootstrap ===\n");
    
    // Check dependencies
    deps::warn_dependencies();
    
    // Ensure directories exist
    paths::ensure_directories()?;
    
    println!("\n✓ Bootstrap complete!");
    println!("\nZircon is installed at: {}", paths::zircon_root().display());
    
    #[cfg(windows)]
    {
        println!("\nNext steps:");
        println!("  1. Add Zircon to your PATH:");
        println!("     PowerShell: $env:Path = \"{};$env:Path\"", paths::bin_dir().display());
        println!("     CMD:        set PATH={};%PATH%", paths::bin_dir().display());
        println!("\n  2. Then load the environment with:");
        println!("     PowerShell: iex (zircon env --shell powershell)");
        println!("     CMD:        zircon env --shell cmd");
        println!("\n  3. Install a zrc version:");
        println!("     zircon build main");
        println!("     zircon build v0.1.0");
    }
    
    #[cfg(not(windows))]
    {
        println!("\nNext steps:");
        println!("  1. Add Zircon to your PATH:");
        println!("     export PATH=\"{}:$PATH\"", paths::bin_dir().display());
        println!("\n  2. Then load the environment with:");
        println!("     source <(zircon env)");
        println!("\n  3. Install a zrc version:");
        println!("     zircon build main");
        println!("     zircon build v0.1.0");
        println!("\n  To make these settings permanent, add to your shell profile (~/.bashrc, ~/.zshrc, etc.):");
        println!("     echo 'source <({}/zircon env)' >> ~/.bashrc", paths::bin_dir().display());
    }
    
    Ok(())
}

File: src/cmds/internal_cmds.c
#include "internal_cmds.h"
#include "../deps.h"
#include "../paths.h"

static int cmd_bootstrap(void) {
    printf("=== Zircon Bootstrap ===\n\n");
    
    if (check_dependencies_strict() != 0) {
        return ERR_GENERAL;
    }
    
    if (ensure_directories() != 0) {
        fprintf(stderr, "Failed to create directories\n");
        return ERR_IO;
    }
    
    printf("\n✓ Bootstrap complete!\n");
    
    char* root = zircon_root();
    printf("\nZircon is installed at: %s\n", root ? root : "(unknown)");
    SAFE_FREE(root);
    
    char* bin = bin_dir();
    printf("\nNext steps:\n");
    printf("  1. Add Zircon to your PATH:\n");
    printf("     export PATH=\"%s:$PATH\"\n", bin ? bin : "(unknown)");
    printf("\n  2. Then load the environment with:\n");
    printf("     source <(zircon env)\n");
    printf("\n  3. Install a zrc version:\n");
    printf("     zircon build main\n");
    printf("     zircon build v0.1.0\n");
    printf("\n  To make these settings permanent, add to your shell profile (~/.bashrc, ~/.zshrc, etc.):\n");
    printf("     echo 'source <(%s/zircon env)' >> ~/.bashrc\n", bin ? bin : "(unknown)");
    SAFE_FREE(bin);
    
    return ERR_OK;
}

int dispatch_internal_command(cli_context_t* ctx) {
    switch (ctx->data.internal_cmd.subcmd) {
        case INTERNAL_CMD_BOOTSTRAP:
            return cmd_bootstrap();
        default:
            return ERR_INVALID_ARG;
    }
}
